tests/unit_tests/verRctNonSemanticsSimple.cpp

```diff
@@ -0,0 +1,287 @@
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <sstream>
+
+#define IN_UNIT_TESTS // to access Blockchain::expand_transaction_2
+
+#include "gtest/gtest.h"
+#include "unit_tests_utils.h"
+
+#include "cryptonote_basic/cryptonote_format_utils.h"
+#include "cryptonote_core/blockchain.h"
+#include "file_io_utils.h"
+#include "misc_log_ex.h"
+#include "ringct/rctSigs.h"
+
+/**
+ * @brief Used in test "cached_serialize_assumption" to test the do_serialize() assumption.
+ *
+ * The purpose of this macro is to assert that changes in rctSig's in memory correlate to changes in
+ * the serialized output of that rctSig. This assumes the existence of scoped variables
+ * "original_sig" and "original_sig_hex".
+ *
+ * @param fieldmodifyclause an expression which modifies a rctSig field
+ */
+#define CACHED_SERIALIZATION_ASSUMPTION_SUBTEST(fieldmodifyclause)       \
+    do {                                                                 \
+        rct::rctSig modded_sig = original_sig;                           \
+        modded_sig.fieldmodifyclause;                                    \
+        const std::string mod_sig_hex = rct_sig_to_hex_hash(modded_sig); \
+        EXPECT_NE(original_sig_hex, mod_sig_hex);                        \
+    } while (0);                                                         \
+
+/**
+ * @brief Used in "cached_is_noncached" to test for cache robustness against key image modifications
+ *
+ * The purpose of this macro is to assert that changes in key images, while not necessarily changing
+ * serialization of rctSig's, will still cause verRctNonSemanticsSimpleCached to fail.
+ *
+ * @param modifyf Takes a cryptonote::transaction& and modifies it
+ */
+#define CACHED_IS_NON_CACHED_SUBTEST(og_tx, input_pubkeys, modifyf)                          \
+    do {                                                                                     \
+        EXPECT_TRUE(rct::verRctNonSemanticsSimple(og_tx.rct_signatures));                    \
+        EXPECT_TRUE(rct::verRctNonSemanticsSimpleCached(og_tx));                             \
```

Comment:
`verRctNonSemanticsSimpleCached` goes through different code paths the first and subsequent times it encounters a tx, so it should be called two times in a row everywhere to be sure.

Reply:
done